rfc-capi-for-kvm-capk/README.md

```diff
@@ -0,0 +1,449 @@
+  don't need to maintain legacy services and libraries (e.g. `k8scloudconfig`, `apiextensions`, `cert-operator`,
+  `cluster-service`) for the foreseeable future.
+- As a GS PE, I want the operator(s) reconciling KVM clusters (collectively referred to as "KVM operators") to work 
+  independently of other GS components so I can add features more quickly and make usage by the wider community more
+  feasible/possible.
+- As a GS PE, I want KVM operators to work in arbitrary Kubernetes clusters so we can stop maintaining our custom
+  provisioning tooling for on-prem management clusters.
+- As a GS employee, I want KVM operators to implement the CAPI provider contract so that we can be in the list of
+  CAPI providers to improve the reputation of our company.
+- As a GS on-premises (on-prem) customer, I want KVM operator to implement CAPI so that I have access to new
+  provider-independent features developed by other GS teams.
+- As a GS oncall engineer, I want on-prem installations to be as similar as possible to the other providers so that I
+  can share best practices, on call routines, and documentation with the other teams.
+- As a GS on-prem customer, I want GS to implement a migration strategy from current kvm clusters to CAPI KVM clusters
+  so that I can access the new features without service disruption and workload migration.
+- As a GS on-prem customer, I want GS to implement a break out/migration strategy for on-prem installations so that I
+  can use other CAPI on-prem providers in the future such as CAPV (Cluster API for VSphere (VMWare)) or metal3 (Cluster
+  API for bare metal).
+- As a member of Team Rocket, I would like to have confidence in my changes to the platform through the use of reliable
+  unit, integration, and end-to-end tests.
+
+
+### Rationale
+
+With the current move to full CAPI adoption in Team Firecracker and Team Celestial, Team Rocket would be the only user
+and therefore the only maintainer of the current components and code that would be legacy for the other teams.
+
+This would mean that we would need to put a lot of effort into understanding and owning this code fully in order to be
+able to fix bugs and maintain KVM installations
+in the future.
+
+Instead of being the sole maintainer of a huge amount of legacy code, Team Rocket wants to invest this time in
+implementing CAPI for KVM operator and maintain relative platform parity with the other GS provider teams.
+
+### Goals
+
+- Customers can upgrade existing clusters to CAPK without significant workload downtime.
+- Existing CAPI-compatible tools such as kubectl-gs, happa, GS api, and clusterctl work with CAPK clusters.
+- Team Rocket is responsible for a significantly reduced number of repos and lines of code.
+
+### Non-goals
+
+- Get CAPK accepted as a CNCF project.
+- Support mixed CAPK/CAPV clusters.
+- Support mixed management clusters (e.g. CAPK on AWS).
+- Support `MachineSet` or `MachinePool`.
```

Comment:
Why not?

Reply:
MachinePool is not a primitive that would be there on KVM, so ok to leave out.

This comment made me check out https://cluster-api.sigs.k8s.io/user/concepts.html though and isn't MachineSet just the versioned instantiation of a MachineDeployment (like in Depployments and Replicasets)? Why would we skip that then?

Reply:
To complete the confusion. I'd even say we don't need MachineDeployments but MachinePools instead. The KVM provider runs on top of Kubernetes. So a MachineSet/MachineDeployment would be managing pods. A MachinePool would make our lives easier as it would use a Deployment for the VirtualMachines.


Example on AWS:
 * MachineSet => Manages individual EC2 instances
 * MachinePool => Manages an ASG

The same would be on K8s:
 * MachineSet => Manages individual pods for the VMs
 * MachinePool => Uses a Deployment that manages the pods

Reply:
Oh and btw. We wanted to use a Deployment for the VMs on KVM in the first place. Just because there were some initial problems with init container, network etc we went with managing individual pods later on. Imo this would be a good time to evaluate this again especially if networking is finally different and we can simply use CNI.

Reply:
In the upcoming version of KVM we are already using CNI AFAIK.

Reply:
> MachinePool would make our lives easier as it would use a Deployment for the VirtualMachines.

I assumed a MachineDeployment already would deliver us the features of what we need from Deployments and we could maybe not have the need for another primitive, i.e. the Deployment, in between, but then I'm not deep enough into the CAPI primitives.

Reply:
I thought about this a bit more and I think Timo's point towards MPs might actually be in line with what Thomas was drafting and that's maybe why he left out the MachineSets. If so, then that does make sense.

Reply:
How  I read the concepts `MachineSets` are not supposed to be used directly, but rather something used under the hood by the `MachineDeployment`. With regards to MachinePoos my impression is that the implementation is slightly  more complex than the MachineDeployments, also they are still experimental.  So I think it is fine to stay with MachineDeployments  for the MVP.

Reply:
It's difficult to find much information about these concepts. I agree that managing node pods with a normal `Deployment` sounds ideal in which case the `MachinePool` seems like a better fit: an ASG is similar to a Deployment for KVM.

It's not clear yet how much node-specific configuration we are going to have. If we want to, for example, mount a specific host path on a single worker node, we would need to create the `Machine` directly without a `MachineDeployment`. As long as the nodes are homogenous, a `MachineDeployment` seems like the easiest approach.

For the time being, I will drop this non-goal so we can make the final decision once the node configuration requirements are a little clearer.